test/e2e/annotations/canary.go

```diff
@@ -192,4 +192,47 @@ var _ = framework.IngressNginxDescribe("Annotations - canary", func() {
 			Expect(body).ShouldNot(ContainSubstring("http-svc-canary"))
 		})
 	})
+
+	Context("Single canary Ingress", func() {
+		It("should not use canary as a catch-all server", func() {
+			host := "foo"
+			canaryIngName := fmt.Sprintf("%v-canary", host)
+			annotations := map[string]string{
+				"nginx.ingress.kubernetes.io/canary":           "true",
+				"nginx.ingress.kubernetes.io/canary-by-header": "CanaryByHeader",
+			}
+
+			ing := framework.NewSingleIngress(canaryIngName, "/", "", f.IngressController.Namespace, "http-svc-canary", 80, &annotations)
+			f.EnsureIngress(ing)
+
+			f.WaitForNginxServer("_",
+				func(server string) bool {
+					if !Expect(server).Should(ContainSubstring("server_name _")) {
+						return false
+					}
+					if !Expect(server).Should(ContainSubstring(`set $ingress_name "";`)) {
+						return false
+					}
+					return Expect(server).Should(ContainSubstring(`set $proxy_upstream_name "upstream-default-backend";`))
+				})
+		})
+
+		It("should not use canary with domain as a server", func() {
+			host := "foo"
+			canaryIngName := fmt.Sprintf("%v-canary", host)
+			annotations := map[string]string{
+				"nginx.ingress.kubernetes.io/canary":           "true",
+				"nginx.ingress.kubernetes.io/canary-by-header": "CanaryByHeader",
+			}
+
+			ing := framework.NewSingleIngress(canaryIngName, "/", host, f.IngressController.Namespace, "http-svc-canary", 80, &annotations)
+			f.EnsureIngress(ing)
+
+			time.Sleep(waitForLuaSync)
+
+			f.WaitForNginxConfiguration(func(cfg string) bool {
+				return Expect(cfg).ShouldNot(ContainSubstring("server_name " + host))
+			})
```

Comment:
I wish we had a better way to implement this test. Currently if for some reason there's a delay in sync, the test will pass regardless of your fix above, because `"server_name " + host` does not exist in the config to start with. In case of catch all you can get away with `Expect(server).Should(ContainSubstring(`set $proxy_upstream_name "upstream-default-backend";`))` but here I don't know how to implement this test better.

One thing that comes to mind is after the canary ingress you can create another ingress with host and then assert that 
```
Expect(cfg).Should(ContainSubstring("server_name " + other_host)) && Expect(cfg).ShouldNot(ContainSubstring("server_name " + host))
```

Reply:
> In case of catch all you can get away with Expect(server).Should(ContainSubstring(set $proxy_upstream_name "upstream-default-backend";))

this is actually wrong, the issue exist in the above test as well